File: src/main/ruby/g0001_0100/s0017_letter_combinations_of_a_phone_number/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Backtracking
# #LeetCode_75_Backtracking #Algorithm_II_Day_11_Recursion_Backtracking
# #Udemy_Backtracking/Recursion #Top_Interview_150_Backtracking #Big_O_Time_O(4^n)_Space_O(n)
# #2023_11_17_Time_49_ms_(93.94%)_Space_210.9_MB_(97.73%)

# @param {String} digits
# @return {String[]}
def letter_combinations(digits)
  return [] if digits.empty?

  letters = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
  ans = []
  find_combinations(0, digits, letters, "", ans)
  ans
end

private

def find_combinations(start, nums, letters, curr, ans)
  if curr.length == nums.length
    ans << curr
    return
  end

  (start...nums.length).each do |i|
    n = nums[i].to_i
    letters[n].each_char do |ch|
      find_combinations(i + 1, nums, letters, curr + ch, ans)
    end
  end
end

File: src/main/dart/g0001_0100/s0017_letter_combinations_of_a_phone_number/Solution.dart
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #String #Hash_Table #Backtracking
// #LeetCode_75_Backtracking #Algorithm_II_Day_11_Recursion_Backtracking
// #Udemy_Backtracking/Recursion #Top_Interview_150_Backtracking #Big_O_Time_O(4^n)_Space_O(n)
// #2024_09_30_Time_293_ms_(92.50%)_Space_148.3_MB_(56.25%)

class Solution {
  List<String> letterCombinations(String digits) {
    if (digits.isEmpty) {
      return [];
    }
    final Map<String, String> hashmap = {
      '2': 'abc',
      '3': 'def',
      '4': 'ghi',
      '5': 'jkl',
      '6': 'mno',
      '7': 'pqrs',
      '8': 'tuv',
      '9': 'wxyz',
    };
    List<String> ans = [];
    void backTracking(String digit, int index, List<String> ans) {
      if (digit.length == digits.length) {
        ans.add(digit);
        return;
      }
      for (String character in hashmap[digits[index]]!.split('')) {
        backTracking(digit + character, index + 1, ans);
      }
    }
    backTracking('', 0, ans);
    return ans;
  }
}
